test/functional/deprecate_warning_tests.js

```diff
@@ -0,0 +1,116 @@
+
+  it('node --no-deprecation flag should suppress all deprecation warnings', {
+    metadata: { requires: { node: '>=6.0.0' } },
+    test: function(done) {
+      exec(
+        'node --no-deprecation ./test/tools/deprecate_warning_test_program.js',
+        (err, stdout, stderr) => {
+          expect(err).to.be.null;
+          expect(stdout).to.be.empty;
+          expect(stderr).to.be.empty;
+          done();
+        }
+      );
+    }
+  });
+
+  it('node --trace-deprecation flag should print stack trace to stderr', {
+    metadata: { requires: { node: '>=6.0.0' } },
+    test: function(done) {
+      exec(
+        'node --trace-deprecation ./test/tools/deprecate_warning_test_program.js',
+        (err, stdout, stderr) => {
+          expect(err).to.be.null;
+          expect(stdout).to.be.empty;
+          expect(stderr).to.not.be.empty;
+
+          const split = stderr.split('\n');
+          const warning = split
+            .shift()
+            .split(')')[1]
+            .trim();
+
+          // ensure warning is the first line printed
+          expect(warning).to.equal(
+            'DeprecationWarning: testDeprecationFlags option [maxScan]' + defaultMessage
+          );
+
+          // ensure each following line is from the stack trace, i.e. 'at config.deprecatedOptions.forEach.deprecatedOption'
+          split.pop();
+          split.forEach(s => {
+            expect(s.trim()).to.match(/^at/);
+          });
+
+          done();
+        }
+      );
+    }
+  });
+
+  it('node --throw-deprecation flag should throw error when deprecated function is called', {
+    metadata: { requires: { node: '>=6.0.0' } },
+    test: function(done) {
+      exec(
+        'node --throw-deprecation ./test/tools/deprecate_warning_test_program.js this_arg_should_never_print',
+        (err, stdout, stderr) => {
+          expect(stderr).to.not.be.empty;
+          expect(err).to.not.be.null;
+          expect(err)
+            .to.have.own.property('code')
+            .that.equals(1);
+
+          // ensure stdout is empty, i.e. that the program threw an error before reaching the console.log statement
+          expect(stdout).to.be.empty;
+          done();
+        }
+      );
+    }
+  });
+
+  it('test behavior for classes with an associated logger', function() {
+    const fakeClass = new ClassWithLogger();
+    const logger = fakeClass.getLogger();
+    const stub = sinon.stub(logger, 'warn');
+
+    fakeClass.f({ maxScan: 5, snapshot: true });
+    fakeClass.f({ maxScan: 5, snapshot: true });
+    expect(stub).to.have.been.calledTwice;
+    ensureCalledWith(stub, [
+      'f option [maxScan] is deprecated and will be removed in a later version.',
+      'f option [snapshot] is deprecated and will be removed in a later version.'
+    ]);
+  });
+
+  it('test behavior for classes without an associated logger', function() {
+    const fakeClass = new ClassWithoutLogger();
+
+    fakeClass.f({ maxScan: 5, snapshot: true });
```

Comment:
there are no assertions here. Is something happening?

Reply:
We are using these tests to ensure that our deprecation function won't error if it's called from a class without a getLogger or with getLogger that returns undefined. When we take out the checks to see if getLogger is there or that logger is defined before using it, these tests will error. The fact that they don't error means that they are passing. Is there a better way to do this?

Reply:
I think @daprahamian s point here is that without an assertion we are not _actually_ checking anything, though you may be able to visually determine its working like you think. I think you can stub out either/both `process.emitWarning`/`console.error` for this purpose.